### PR TITLE
Add support for SPI TFT displays and touch screen for STM32H7 MCUs

### LINKED ISSUE #25046: [FR] Support TFT SPI screens on STM32H7 platform
### Is your feature request related to a problem? Please describe.

Currently "[SPI TFT is currently only supported on STM32F1 and STM32F4 hardware](https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.1.x/Marlin/src/HAL/STM32/tft/tft_spi.h#L29)", while new BTT SKR 3 board is STM32H7 based. Thus it's impossible to replace stock BTT SKR 2 board with SKR 3, as Biqu S1 SE uses TFT35 SPI screen.

I tried to add `#include "stm32h7xx_hal.h"`, but build failed with error 

```
Marlin\src\HAL\STM32\tft\xpt2046.cpp:146:18: error: 'struct SPI_TypeDef' has no member named 'DR'; did you mean 'SR'?
  146 |   SPIx.Instance->DR = data;
```
https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.1.x/Marlin/src/HAL/STM32/tft/xpt2046.cpp#L146

### Are you looking for hardware support?

Biqu S1 SE (Plus)

### Describe the feature you want

Support TFT SPI screens on STM32H7 pratform

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jmz52**
### Description

Add support for SPI TFT displays and touch screen for STM32H7 MCUs

### Requirements

Board with STM32H7 MCU and SPI TFT display
 -  BigTreeTech SKR 3 or custom STM32H7 board
 - TFT display with SPI interface

### Benefits

BigTreeTech SKR 3 can be used with TFT displays
- generic SPI TFT module
- MKS Robin TS35
- display from BIQU B1 SE Plus

### Related Issues
https://github.com/MarlinFirmware/Marlin/issues/25046

### CODE REVIEW

**Comment by thisiskeithb** [comment]
`HAS_SPI_TFT` needs to be split up in all the pins files since `BTT_TFT35_SPI_V1_0` and `MKS_TS35_V2_0` do not share the same pinout.

Using an SKR 3 EZ as the test board:
- `SPEAKER` on `MKS_TS35_V2_0` causes the graphics to go haywire when you send a tone with `M300`. Clicking the encoder or touching the TFT fixes it, but I could never get it to stop. There's also an occasional glitch while navigating the menus/UI.
- `SPEAKER` on `BTT_TFT35_SPI_V1_0` works fine.

<details><summary>Test config diff:</summary>
<p>

```diff
diff --git a/Marlin/Configuration.h b/Marlin/Configuration.h
index 8818b04..e82601d 100644
--- a/Marlin/Configuration.h
+++ b/Marlin/Configuration.h
@@ -37,6 +37,10 @@
  */
 #define CONFIGURATION_H_VERSION 02010300
 
+// Define TFTs here for faster testing
+#define MKS_TS35_V2_0
+//#define BTT_TFT35_SPI_V1_0
+
 //===========================================================================
 //============================= Getting Started =============================
 //===========================================================================
@@ -67,7 +71,7 @@
 
 // Choose the name from boards.h that matches your setup
 #ifndef MOTHERBOARD
-  #define MOTHERBOARD BOARD_RAMPS_14_EFB
+  #define MOTHERBOARD BOARD_BTT_SKR_V3_0_EZ
 #endif
 
 /**
@@ -78,7 +82,7 @@
  *
  * :[-1, 0, 1, 2, 3, 4, 5, 6, 7]
  */
-#define SERIAL_PORT 0
+#define SERIAL_PORT -1
 
 /**
  * Serial Port Baud Rate
@@ -2289,13 +2293,13 @@
  *   M501 - Read settings from EEPROM. (i.e., Throw away unsaved changes)
  *   M502 - Revert settings to "factory" defaults. (Follow with M500 to init the EEPROM.)
  */
-//#define EEPROM_SETTINGS     // Persistent storage with M500 and M501
+#define EEPROM_SETTINGS       // Persistent storage with M500 and M501
 //#define DISABLE_M503        // Saves ~2700 bytes of flash. Disable for release!
 #define EEPROM_CHITCHAT       // Give feedback on EEPROM commands. Disable to save PROGMEM.
 #define EEPROM_BOOT_SILENT    // Keep M503 quiet and only give errors during first load
 #if ENABLED(EEPROM_SETTINGS)
-  //#define EEPROM_AUTO_INIT  // Init EEPROM automatically on any errors.
-  //#define EEPROM_INIT_NOW   // Init EEPROM on first boot after a new build.
+  #define EEPROM_AUTO_INIT    // Init EEPROM automatically on any errors.
+  #define EEPROM_INIT_NOW     // Init EEPROM on first boot after a new build.
 #endif
 
 // @section host
@@ -2668,7 +2672,7 @@
 // If you have a speaker that can produce tones, enable it here.
 // By default Marlin assumes you have a buzzer with a fixed frequency.
 //
-//#define SPEAKER
+#define SPEAKER
 
 //
 // The duration and frequency for the UI feedback sound.
@@ -3272,7 +3276,7 @@
  *   root of your SD card, together with the compiled firmware.
  */
 //#define TFT_CLASSIC_UI
-//#define TFT_COLOR_UI
+#define TFT_COLOR_UI
 //#define TFT_LVGL_UI
 
 #if ENABLED(TFT_COLOR_UI)
@@ -3318,7 +3322,7 @@
 //
 // Touch Screen Settings
 //
-//#define TOUCH_SCREEN
+#define TOUCH_SCREEN
 #if ENABLED(TOUCH_SCREEN)
   #define BUTTON_DELAY_EDIT      50 // (ms) Button repeat delay for edit screens
   #define BUTTON_DELAY_MENU     250 // (ms) Button repeat delay for menus
diff --git a/platformio.ini b/platformio.ini
index e3bdb6f..12f2fc4 100644
--- a/platformio.ini
+++ b/platformio.ini
@@ -13,7 +13,7 @@
 [platformio]
 src_dir      = Marlin
 boards_dir   = buildroot/share/PlatformIO/boards
-default_envs = mega2560
+default_envs = STM32H723Vx_btt
 include_dir  = Marlin
 extra_configs =
     Marlin/config.ini
```

</p>
</details>

**Comment by thisiskeithb** [comment]
I just rebased this PR to current bugfix & pushed [`afdcc19`](https://github.com/MarlinFirmware/Marlin/pull/25784/commits/afdcc1935ce7ba1ae4dda2474deb59cc1ad58b8b) which splits up `HAS_SPI_TFT` in `pins_BTT_SKR_V3_0_common.h` to support both `BTT_TFT35_SPI_V1_0` & `MKS_TS35_V2_0` since they do not share the same pinout.

Similar changes will need be done to the rest of the pins files as mentioned in my previous comment, so if this is OK, I can proceed with updating the rest of the pins files.

File: Marlin/src/pins/stm32h7/pins_BTT_SKR_V3_0_common.h [hunk 1]
```diff
@@ -458,6 +458,98 @@
...
+    #ifndef TFT_WIDTH
+      #define TFT_WIDTH                      480
+    #endif
+    #ifndef TFT_HEIGHT
+      #define TFT_HEIGHT                     320
+    #endif
+
+    #if ENABLED(BTT_TFT35_SPI_V1_0)
+      // 480x320, 3.5", SPI Display with Rotary Encoder.
+      // Stock Display for the BIQU B1 SE.
```
Comment (thisiskeithb) [inline Marlin/src/pins/stm32h7/pins_BTT_SKR_V3_0_common.h:478]: This should be Biqu B1 SE & SE Plus (or just B1 SE Series?)

**Comment by thisiskeithb** [comment]
This PR has been merged, but the later changes no longer compile. ☹️

**Comment by jmz52** [comment]
@thisiskeithb, can you provide me configuration files you are using? I don't have access to my H7 board or its build environment at the moment.

**Comment by thisiskeithb** [comment]
I am putting together a PR right now since this broke compiling on the SKR V2 as well.

**Comment by thisiskeithb** [comment]
SKR V3 + SPI TFTs are fixed in https://github.com/MarlinFirmware/Marlin/pull/26041, but I'm still looking into why the SKR 2 builds are broken now.

**Comment by thisiskeithb** [comment]
@jmz52: Trying out the original commit before everything was rebased / updated (https://github.com/MarlinFirmware/Marlin/tree/b28fa1955a401ddedd39bc462804691cb9fdf2f8), I still get a white screen with a `BTT_TFT35_SPI_V1_0` on `BOARD_BTT_SKR_V2_0_REV_B` (STM32F4).

For testing/reference, you can use https://github.com/MarlinFirmware/Configurations/tree/bugfix-2.1.x/config/examples/BIQU/B1%20SE%20Plus

**Comment by thinkyhead** [comment]
What article, datasheet, or documentation is best for understanding these changes, and what is the current theory on why the applied changes are not working?

**Comment by jmz52** [comment]
Fixed in linked PR
Problem was in `TFT_SPI::transmit()` function
First issue was with SPI busy check immediately after data is send to data register.
```
   SPIx.Instance->DR = data;
    while (__HAL_SPI_GET_FLAG(&SPIx, SPI_FLAG_BSY)) {}
```
SPI status is reported as idle because data transfer is not actually started yet.
Added an extra check to wait for data transfer to actually start before waiting for data transfer completion.
```
while (__HAL_SPI_GET_FLAG(&SPIx, SPI_FLAG_BSY)) {}
while (!__HAL_SPI_GET_FLAG(&SPIx, SPI_FLAG_TXE)) {}   // Wait for data transfer to actually start
while (__HAL_SPI_GET_FLAG(&SPIx, SPI_FLAG_BSY)) {}    // Wait until SPI is idle
```

Second issue was with `__HAL_SPI_DISABLE(&SPIx);` after data transfer is completed.
This call it is required for STM32H7 to allow modification of `SPIx.Instance->CR2` register.
However of STM32F4 it seems to mess with SPI SCK line corrupting data transfer altogether.
This can be mitigation be toggling `TFT_CS_PIN`, but it is easier to limit this call to H7 code only.

Code works just fine on my MKS Robin Nano v3.1 board.

**Comment by jmz52** [comment]
Please note that Color_UI layout is a mess.
I'll going to clean the layout, but I need a few days to get this sorted out.

Following issues were observed while testing F4 fix:

320x240
- SD card icon is flickering when card is not inserted

240x320
- Two SD card icons, one always white (like card is inserted)
- Two Menu icons, one is off the screen
- Print time misaligned, it overlaps with progress bar and SD card icon

480x320
- SD card and Menu icons are not displayed making it impossible to enter menu without encoder
- Navigation icons misaligned

320x480
- SD card and Menu icons are not displayed making it impossible to enter menu without encoder
- Navigation icons misaligned
- Fan icon misaligned
- Coordinates bar is too small for XYZ coordinates

480x272
- SD card and Menu icons are not displayed making it impossible to enter menu without encoder
- Navigation icons misaligned
- Heaters and Fans areas overlaps with Coordinates bar, causing flickering

272x480
- Not implemented, should throw error message for this configuration

**Comment by ellensp** [comment]
@jmz52 https://github.com/MarlinFirmware/Marlin/pull/24843#issuecomment-1627226620
480x320 status screen is missing the settings icon too.

**Comment by lukasradek** [comment]
@jmz52 
That will be greatly appreciated!

**Comment by vovodroid** [comment]
> still get a white screen with a `BTT_TFT35_SPI_V1_0` on `BOARD_BTT_SKR_V2_0_REV_B` (STM32F4).

That's why I got today white screen on boot after flashing latest bugfix-2.1.x branch to Biqu B1 SE with SKR 2 board.

**Comment by thisiskeithb** [comment]
> That's why I got today white screen on boot after flashing latest bugfix-2.1.x branch to Biqu B1 SE with SKR 2 board.

See https://github.com/MarlinFirmware/Marlin/pull/26052

**Comment by vovodroid** [comment]
Hi,

what is maturity of this PR?

It seems that something wrong with my SKR 2, and as it's not available, probably I need to replace it with SKR 3. Does it work properly with SPI TFT in B1 SE/Plus?

Thanks.

**Comment by thisiskeithb** [comment]
> what is maturity of this PR?

This PR has been merged. Follow https://github.com/MarlinFirmware/Marlin/pull/26052 for the current state of SPI TFT support.

**Comment by vovodroid** [comment]
**Fix SPI TFT for STM32F4 #26052** about SKR 2, but I would like to know whether there are any known problems with STM32H7, for example SKR 3.

**Comment by thisiskeithb** [comment]
> I would like to know whether there are any known problems with STM32H7, for example SKR 3.

You have to add `#define TFT_DRIVER ST7796` to your config, but `BTT_TFT35_SPI_V1_0` works on `BOARD_BTT_SKR_V3_0`/`BOARD_BTT_SKR_V3_0_EZ` with current `bugfix-2.1.x` (19869d3).

**Comment by vovodroid** [comment]
First, thanks a lot for supporting TFT SPI on SKR 3! My SKR 2 died and was replaced by SKR 3.

But I also see that though some related changes were done for LPC1768 platform it still can not be build for SKR 1.3/1.4, it fails with **#error "Invalid TFT SPI configuration."** in tft_spi.h:

```
#define IS_SPI(N) (BOARD_NR_SPI >= N && (TFT_SCK_PIN == BOARD_SPI##N##_SCK_PIN) && (TFT_MOSI_PIN == BOARD_SPI##N##_MOSI_PIN) && (TFT_MISO_PIN == BOARD_SPI##N##_MISO_PIN))
#if IS_SPI(1)
  #define TFT_SPI_DEVICE  1
  #define LPC_SSPx  LPC_SSP0
#elif IS_SPI(2)
  #define TFT_SPI_DEVICE  2
  #define LPC_SSPx  LPC_SSP1
#else
 #error "Invalid TFT SPI configuration."
#endif
#undef IS_SPI
```

**Comment by ellensp** [comment]
@vovodroid LPC1768/LPC1769 fixes  are in https://github.com/MarlinFirmware/Marlin/pull/26176

**Comment by DragonLord1981** [comment]
Any updates on this mather?

Motherboard MKS_SGEN_L_V2
Screen MKS TS35 V2.0

I get this error 

Compiling .pio\build\LPC1769\src\src\HAL\LPC1768\usb_serial.cpp.o
In file included from Marlin\src\HAL\LPC1768\tft\tft_spi.cpp:29:
Marlin\src\HAL\LPC1768\tft\tft_spi.h:38:4: error: #error "Invalid TFT SPI configuration."
   38 |   #error "Invalid TFT SPI configuration."
      |    ^~~~~
Compiling .pio\build\LPC1769\src\src\HAL\STM32F1\tft\tft_spi.cpp.o
Compiling .pio\build\LPC1769\src\src\HAL\STM32F1\tft\xpt2046.cpp.o
Compiling .pio\build\LPC1769\src\src\HAL\STM32\tft\tft_spi.cpp.o
Compiling .pio\build\LPC1769\src\src\HAL\STM32\tft\xpt2046.cpp.o
Compiling .pio\build\LPC1769\src\src\HAL\shared\Delay.cpp.o
Compiling .pio\build\LPC1769\src\src\HAL\shared\HAL.cpp.o
Compiling .pio\build\LPC1769\src\src\HAL\shared\MinSerial.cpp.o
Compiling .pio\build\LPC1769\src\src\HAL\shared\backtrace\backtrace.cpp.o
Marlin\src\HAL\LPC1768\tft\tft_spi.h: In static member function 'static void TFT_SPI::DataTransferEnd()':
Marlin\src\HAL\LPC1768\tft\tft_spi.h:73:68: error: 'LPC_SSPx' was not declared in this scope; did you mean 'LPC_SSP0'?
   73 |   static void DataTransferEnd() { WRITE(TFT_CS_PIN, HIGH); SSP_Cmd(LPC_SSPx, DISABLE); };       
      |                                                                    ^~~~~~~~
      |                                                                    LPC_SSP0
Marlin\src\HAL\LPC1768\tft\tft_spi.cpp: At global scope:
Marlin\src\HAL\LPC1768\tft\tft_spi.cpp:31:24: error: 'TFT_SPI_DEVICE' was not declared in this scope    
   31 | SPIClass TFT_SPI::SPIx(TFT_SPI_DEVICE);
      |                        ^~~~~~~~~~~~~~
Marlin\src\HAL\LPC1768\tft\tft_spi.cpp: In static member function 'static void TFT_SPI::Init()':
Marlin\src\HAL\LPC1768\tft\tft_spi.cpp:46:18: error: 'TFT_SPI_DEVICE' was not declared in this scope    
   46 |   SPIx.setModule(TFT_SPI_DEVICE);
Compiling .pio\build\LPC1769\src\src\HAL\shared\backtrace\unwmemaccess.cpp.o
      |                  ^~~~~~~~~~~~~~
Marlin\src\HAL\LPC1768\tft\tft_spi.cpp: In static member function 'static bool TFT_SPI::isBusy()':      
Marlin\src\HAL\LPC1768\tft\tft_spi.cpp:103:24: error: 'LPC_SSPx' was not declared in this scope; did you mean 'LPC_SSP0'?
  103 |     if ((SSP_GetStatus(LPC_SSPx, SSP_STAT_TXFIFO_EMPTY) == RESET) || (SSP_GetStatus(LPC_SSPx, SSP_STAT_BUSY) == SET)) return true;
      |                        ^~~~~~~~
      |                        LPC_SSP0
compilation terminated due to -fmax-errors=5.
*** [.pio\build\LPC1769\src\src\HAL\LPC1768\tft\tft_spi.cpp.o] Error 1
====================================== [FAILED] Took 3.81 seconds ======================================
Environment    Status    Duration
-------------  --------  ------------
LPC1769        FAILED    00:00:03.808
================================ 1 failed, 0 succeeded in 00:00:03.808 ================================ 

I dident have this problem whit the older marlin 2.1.2 but whit 2.1.2.3 i have it ??

**Comment by thisiskeithb** [comment]
@DragonLord1981: Please [file a proper Bug Report](https://github.com/MarlinFirmware/Marlin/issues/new/choose) and be sure to fill out the Issue template completely & include your `bugfix-2.1.x` configs.

**Pull request merged**